packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/TaskService.java

```diff
@@ -0,0 +1,704 @@
+
+    if (taskManager == null) {
+      weakRef = sHeadlessTaskManagers.get(appId);
+    }
+    return weakRef == null ? null : weakRef.get();
+  }
+
+  private Map<String, Object> exportTaskToHashmap(TaskInterface task) {
+    Map<String, Object> map = new HashMap<>();
+    String consumerClassName = unversionedClassNameForClass(task.getConsumer().getClass());
+
+    map.put("name", task.getName());
+    map.put("consumerClass", consumerClassName);
+    map.put("options", task.getOptions());
+
+    return map;
+  }
+
+  private AppLoaderInterface createAppLoader() {
+    // for now only react-native apps in Expo are supported
+    return AppLoaderProvider.createLoader("react-native-experience", mContext);
+  }
+
+  private boolean loadApp(final String appId, String appUrl) {
+    AppLoaderInterface appLoader = createAppLoader();
+
+    if (appLoader == null) {
+      Log.e(TAG, "Cannot execute background task because application loader can't be found.");
+      return false;
+    }
+    if (appUrl == null) {
+      Log.e(TAG, "Cannot execute background task because application URL is invalid: " + appUrl);
+      return false;
+    }
+
+    // TODO(@tsapeta): add timeout option;
+    Map<String, Object> options = new HashMap<>();
+
+    Log.i(TAG, "Loading headless app '" + appId + "' with url '" + appUrl + "'.");
+
+    AppRecordInterface appRecord = appLoader.loadApp(appUrl, options, new AppLoaderProvider.Callback() {
+      @Override
+      public void onComplete(boolean success, Error error) {
+        if (error != null) {
+          error.printStackTrace();
+          Log.e(TAG, error.getMessage());
+        }
+        if (!success) {
+          sEvents.remove(appId);
+          sEventsQueues.remove(appId);
+          sAppRecords.remove(appId);
+
+          // Host unreachable? Unregister all tasks for that app.
+          unregisterAllTasksForAppId(appId);
+        }
+      }
+    });
+
+    sAppRecords.put(appId, appRecord);
+    return true;
+  }
+
+  private void invalidateAppRecord(String appId) {
+    AppRecordInterface appRecord = sAppRecords.get(appId);
+
+    if (appRecord != null) {
+      appRecord.invalidate();
+      sAppRecords.remove(appId);
+      sHeadlessTaskManagers.remove(appId);
+      Log.i(TAG, "Invalidated headless app '" + appId + "'.");
+    }
+  }
+
+  private void finishJobAfterTimeout(final JobService jobService, final JobParameters params, long timeout) {
+    new Timer().schedule(new TimerTask() {
+      @Override
+      public void run() {
+        jobService.jobFinished(params, false);
+      }
+    }, timeout);
+  }
+
+  public static Map<String, Object> jsonToMap(String jsonStr, boolean recursive) {
+    try {
+      return jsonToMap(new JSONObject(jsonStr), recursive);
+    } catch (JSONException e) {
+      return new HashMap<>();
+    }
+  }
+
+  private static Map<String, Object> jsonToMap(JSONObject json, boolean recursive) {
```

Comment:
I think those static methods could go into a separate utils/helpers class, but if you want to leave them here I won't argue 🙂